draft-ietf-acme-acme.md

```diff
@@ -1000,6 +1000,120 @@ name validation.
 }
 ~~~~~~~~~~
 
+### Status Changes
+
+Each ACME object type goes through a simple state machine over its
+lifetime.  The "status" field of the object indicates which state
+the object is currently in.
+
+Challenge objects are created in the "pending" state.  They
+transition to the "processing" state when the client responds to the
+challenge (see {{responding-to-challenges}}) and the server begins
+attempting to validate that the client has completed the challenge.
+Note that within the "processing" state, the server may attempt to
+validate the challenge multiple times (see {{retrying-challenges}}).
+Likewise, client requests for retries do not cause a state change.
+If validation is successful, the challenge moves to the "valid"
+state; if there is an error, the challenge moves to the "invalid"
```

Comment:
If there is an error, and the challenge moves to the "invalid" stat, how does the client request retry? That seems like the primary reason you would want to retry and there isn't any information about how the server should handle failures during processing except to change out of processing.

Reply:
I think "invalid" is a final decision. If the client wants another try, they'd need to create a new order.
Section 8.2 says that "the server SHOULD retry the query after some time".
Being unable to retry individual challenges might mean that some users will have a hard time acquiring a certificate. But then they should just fix their setup and respond to the challenge only when it will work.
I.e. I agree with the above change.

Reply:
Yeah, what @joernheissler said.  I had also been thinking of "invalid" as a final state.  ISTM if you want to signal to the client that something has gone wrong and a retry is needed, there are basically three options:

1. Keep it in "processing" and add to the subproblems array
2. Move it to "invalid" and add to the subproblems array
3. Move it to a new "retry" state and add to the subproblems array

(1) is what we do now, and in case you can't tell by my phrasing here, I don't think we need any more signal than adding to the subproblems array.  It seems like the only real argument for having a non-"processing" state is if find it distasteful to have "processing" with "error" populated, in which case I would propose we do (3) instead.

Reply:
Ok. That sounds reasonable.